drivers/ng_at86rf2xx/ng_at86rf2xx.c

```diff
@@ -0,0 +1,237 @@
+    memcpy(&addr_long, cpudid, 8);
+    ng_at86rf2xx_set_addr_long(dev, addr_long);
+    /* now compress the long addr to form the short address */
+    for (int i = 2; i < 8; i++) {
+        cpudid[i & 0x01] ^= cpudid[i];
+    }
+    memcpy(&addr_short, cpudid, 2);
+    ng_at86rf2xx_set_addr_short(dev, addr_short);
+#else
+    ng_at86rf2xx_set_addr_long(dev, NG_AT86RF2XX_DEFAULT_ADDR_LONG);
+    ng_at86rf2xx_set_addr_short(dev, NG_AT86RF2XX_DEFAULT_ADDR_SHORT);
+#endif
+    /* set default PAN id */
+    ng_at86rf2xx_set_pan(dev, NG_AT86RF2XX_DEFAULT_PANID);
+    /* set default channel */
+    ng_at86rf2xx_set_chan(dev, NG_AT86RF2XX_DEFAULT_CHANNEL);
+    /* set default TX power */
+    ng_at86rf2xx_set_txpower(dev, NG_AT86RF2XX_DEFAULT_TXPOWER);
+    /* set default options */
+    ng_at86rf2xx_set_option(dev, OPT_AUTOACK, true);
+    ng_at86rf2xx_set_option(dev, OPT_CSMA, true);
+    ng_at86rf2xx_set_option(dev, OPT_TELL_RX_END, true);
+    /* set default protocol */
+#ifdef MODULE_NG_SIXLOWPAN
+    dev->proto = NG_NETTYPE_SIXLOWPAN;
+#else
+    dev->proto = NG_NETTYPE_UNDEF;
+#endif
+    /* enable safe mode (protect RX FIFO until reading data starts) */
+    tmp = NG_AT86RF2XX_TRX_CTRL_2_MASK__RX_SAFE_MODE;
+#ifdef MODULE_NG_AT86RF212
+    /* settings used by Linux 4.0rc at86rf212b driver */
+    tmp |= (NG_AT86RF2XX_TRX_CTRL_2_MASK__SUB_MODE
+            | NG_AT86RF2XX_TRX_CTRL_2_MASK__OQPSK_SCRAM_EN);
+#endif
+    ng_at86rf2xx_reg_write(dev, NG_AT86RF2XX_REG__TRX_CTRL_2, tmp);
+    /* enable interrupts */
+    ng_at86rf2xx_reg_write(dev, NG_AT86RF2XX_REG__IRQ_MASK,
+                           (NG_AT86RF2XX_IRQ_STATUS_MASK__RX_START |
+                            NG_AT86RF2XX_IRQ_STATUS_MASK__TRX_END));
+    /* go into RX state */
+    ng_at86rf2xx_set_state(dev, STATE_RX_AACK_ON);
+
+    DEBUG("ng_at86rf2xx_reset(): reset complete.\n");
+}
+
+bool ng_at86rf2xx_cca(ng_at86rf2xx_t *dev)
+{
+    uint8_t tmp;
+    uint8_t status;
+
+    /* trigger CCA measurment */
+    tmp = ng_at86rf2xx_reg_read(dev, NG_AT86RF2XX_REG__PHY_CC_CCA);
+    tmp &= NG_AT86RF2XX_PHY_CC_CCA_MASK__CCA_REQUEST;
+    ng_at86rf2xx_reg_write(dev, NG_AT86RF2XX_REG__PHY_CC_CCA, tmp);
+    /* wait for result to be ready */
+    do {
+        status = ng_at86rf2xx_reg_read(dev, NG_AT86RF2XX_REG__TRX_STATUS);
+    } while (!(status & NG_AT86RF2XX_TRX_STATUS_MASK__CCA_DONE));
+    /* return according to measurement */
+    if (status & NG_AT86RF2XX_TRX_STATUS_MASK__CCA_STATUS) {
+        return true;
+    }
+    else {
+        return false;
+    }
+}
+
+size_t ng_at86rf2xx_send(ng_at86rf2xx_t *dev, uint8_t *data, size_t len)
+{
+    /* check data length */
+    if (len > NG_AT86RF2XX_MAX_PKT_LENGTH) {
+        DEBUG("[ng_at86rf2xx] Error: data to send exceeds max packet size\n");
+        return 0;
+    }
+    ng_at86rf2xx_tx_prepare(dev);
+    ng_at86rf2xx_tx_load(dev, data, len, 0);
+    ng_at86rf2xx_tx_exec(dev);
```

Comment:
Shouldn´t we move the ```ng_at86rf2xx_tx_exec(dev);``` part to a seperate function to allow preloading? In the background of timing constraints it might be useful to avoid the distortion caused by ```ng_at86rf2xx_tx_load(dev, data, len, 0);``` execution time (long SPI transfer).